lib/Sema/CSDiagnostics.cpp

```diff
@@ -188,3 +189,44 @@ bool LabelingFailure::diagnose() {
                                     CorrectLabels,
                                     isa<SubscriptExpr>(call->getFn()));
 }
+
+bool RValueTreatedAsLValueFailure::diagnose() {
+  if (auto callExpr = dyn_cast<ApplyExpr>(getAnchor())) {
+    Expr *argExpr = callExpr->getArg();
+
+    Diag<StringRef> subElementDiagID;
+    Diag<Type> rvalueDiagID;
+    Expr *diagExpr = nullptr;
+
+    if (isa<PrefixUnaryExpr>(callExpr) || isa<PostfixUnaryExpr>(callExpr)) {
+      subElementDiagID = diag::cannot_apply_lvalue_unop_to_subelement;
+      rvalueDiagID = diag::cannot_apply_lvalue_unop_to_rvalue;
+      diagExpr = argExpr;
+    } else if (isa<BinaryExpr>(callExpr)) {
+      subElementDiagID = diag::cannot_apply_lvalue_binop_to_subelement;
+      rvalueDiagID = diag::cannot_apply_lvalue_binop_to_rvalue;
+      auto argTuple = dyn_cast<TupleExpr>(argExpr);
+      diagExpr = argTuple->getElement(0);
+    } else {
+      auto lastPathElement = getLocator()->getPath().back();
+      assert(lastPathElement.getKind() == ConstraintLocator::PathElementKind::ApplyArgToParam);
+
+      subElementDiagID = diag::cannot_pass_rvalue_inout_subelement;
+      rvalueDiagID = diag::cannot_pass_rvalue_inout;
+      if (auto argTuple = dyn_cast<TupleExpr>(argExpr))
+        diagExpr = argTuple->getElement(lastPathElement.getValue());
+      else if (auto parens = dyn_cast<ParenExpr>(argExpr))
+        diagExpr = parens->getSubExpr();
+    }
+
+    // FIXME: Would like for this to be unnecessary.
```

Comment:
Right, I guess diagnoseSubElementFailure() looks at the type checked expression with a solution applied to figure out what it is. Once we can piece that together from the constraint system itself with locators we should be able to rework this part (and make diagnoseSubElementFailure() static again).

Reply:
Yeah, digging in further it looks like the goal here is to, given “foo.bar”, figure out if the real problem is that bar is get-only, or if it’s settable, but “foo” is an rvalue. @xedin how do you envision this working in the fullness of time? Would a mutating member use on an rvalue create a disjunction where one branch has a fix to turn the rvalue into an lvalue?

Reply:
Yes, I assume that we'll have to do something like (what you describe) - and I'm not sure if that's possible - in `shouldAttemptFixes` mode we'd lookup both mutating and non-mutating methods and if base was rvalue but lookup returned mutating method, it would record a fix in the choice.

Reply:
So I looked into this more, and at least here, none of that ought to be necessary, hopefully. The only problem is that `diagnoseSubElementFailure()` is only getting passed the `CS` and `expr` and can only pull out decl bindings from those, and the required overload choices are only in the `Solution`.

So it ought to be fine once `diagnoseSubElementFailure()` gets changed a bit to deal with the `Solution` rather than the `CS` directly.